pyro/infer/mcmc/util.py

```diff
@@ -153,3 +132,81 @@ def log_prob(self, model_trace):
                 return model_trace.log_prob_sum()
             self._compute_log_prob_terms(model_trace)
             return self._aggregate_log_probs(ordinal=frozenset()).sum()
+
+
+class TraceEinsumEvaluator(object):
+    """
+    Computes the log probability density of a trace (of a model with
+    tree structure) that possibly contains discrete sample sites
+    enumerated in parallel. This uses optimized `einsum` operations
+    to marginalize out the the enumerated dimensions in the trace
+    via :class:`~pyro.infer.contract.contract_to_tensor`.
+
+    :param model_trace: execution trace from a static model.
+    :param bool has_enumerable_sites: whether the trace contains any
+        discrete enumerable sites.
+    :param int max_iarange_nesting: Optional bound on max number of nested
+        :func:`pyro.iarange` contexts.
+    """
+    def __init__(self,
+                 model_trace,
+                 has_enumerable_sites=False,
+                 max_iarange_nesting=float("inf")):
+        self.has_enumerable_sites = has_enumerable_sites
+        self.max_iarange_nesting = max_iarange_nesting
+        # To be populated using the model trace once.
+        self._enum_dims = {}
+        self.ordering = {}
+        self._populate_cache(model_trace)
+
+    def _populate_cache(self, model_trace):
+        """
+        Populate the ordinals (set of ``CondIndepStack`` frames)
+        and enum_dims for each sample site.
+        """
+        if not self.has_enumerable_sites:
+            return
+        if self.max_iarange_nesting == float("inf"):
+            raise ValueError("Finite value required for `max_iarange_nesting` when model "
+                             "has discrete (enumerable) sites.")
+        model_trace.compute_log_prob()
+        self.ordering = {name: frozenset(site["cond_indep_stack"])
```

Comment:
I believe ordinals should only contain vectorized frames:
```py
frozenset(f for f in site["cond_indep_stack"] if f.vectorized)
```

Reply:
Done; in HMC, there shouldn't be non-vectorized frames though.